### PR TITLE
[BugFix][KDA] Allocate Post-WU workspace for padded chunks

### PR DESCRIPTION
**Pull request by maoxx241**
### What this PR does / why we need it?

`chunk_kda_fwd` allocates Post-WU scratch using the packed token count, but `WScratchOffset` addresses it using `total_chunks * chunk_size`. When sequences contain tails, full-chunk Cube writes can overrun this allocation and corrupt the U seed buffer that is still in use. This produces NaN outputs/states and can lead to all-invalid speculative sampling rows (`valid_sampled_tokens_count=0`).

For example, 127 sequences alternating between 16, 32 and 64 tokens contain 4,720 tokens. With six heads and K=128, the old allocation is 14,499,840 bytes, while the padded layout requires 24,969,216 bytes. Allocate scratch using the padded chunk count, matching both the generic and A5 addressing. This also covers fixed-length inputs with a partial final chunk.

Add regression coverage for 127/128 variable-length sequences and fixed lengths 80/129, with both recompute modes, comparing outputs and final states against an independent CPU reference.

### Does this PR introduce _any_ user-facing change?

Fixes incorrect KDA outputs and downstream sampling errors for affected sequence batches. No API or configuration change.

### How was this patch tested?

On Ascend A3 with CANN 9.1.0:

- Replayed the same saved failing operator inputs twice before and after rebuilding only the tiling library. Before: U contained 22,612 NaNs and 316 Infs. After: all 12 returned tensors were finite; inputs were unchanged.
- New regression matrix: 8 passed. Restoring the old library makes `[127-None-True]` fail; the fixed library was restored afterward.
- Full `test_chunk_kda_aclnn.py`: 21 passed, 5 A5-specific cases skipped.
- `bash format.sh ci` and `git diff --check` passed.

Full-service validation on a single A3 node uses TP16/EP16, a 16-expert Kimi K3 target, GQA DSpark with 7 speculative tokens, prefix caching disabled, and `max_num_seqs=256`. All three phases reached 128 concurrent running requests and passed: 128 short requests with 256 output tokens, 512 mixed 16/32/64-token prompts with 128 output tokens, and another 256 mixed requests after removing the diagnostic hooks. Total: 896/896 HTTP 200 responses and 131,072 output tokens, with no missing/nonfinite selected logprobs or service errors. Each of the 16 ranks recorded 718 steps and 73,631 request rows during the first two phases: valid counts stayed in 1–8, with zero undiscarded count-zero rows and no unconsumed trace steps. Every phase ended with no running/waiting requests; final health returned 200.

The before/after service runs use the same separate sampler-sort memory fix; the standalone operator tests directly exercise this KDA change. These pruned-weight runs validate the reported count-zero/service-failure behavior, not model accuracy.

The NPU-tested sources are byte-identical to the files in this PR. The release branch advanced from `91b905364` to `9a3a8fdeb` during validation with Docker/CI-only changes.

- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

**Comment by maoxx241** [comment]
## K3 GQA DSpark：decode count=0 的 KDA 根因与修复验证报告

2026-09-09，北京时间。已复现未被 discard 的 decode 行 `valid_sampled_tokens_count=0`，定位并修复 KDA Post-WU 临时区分配不足。修复已通过原失败输入重放、旧库反向对照、独立 CPU 数值参考以及完整服务回归：128 并发下 896/896 请求成功，采集的 16 rank 均未再出现 count=0。PR：[vllm-project/vllm-ascend#16087](https://github.com/vllm-project/vllm-ascend/pull/16087)，目标为 `releases/v0.27.1rc`。

本报告对应当前 PR head `f27f1658669bd22f20e8b2ed539b724ef3bbb5e3`。算子修复提交为 `8ddf72b3b697c29b8b0cd98401dd396fa319fd26`，后续提交只给回归测试补充了参考 final state 的非空断言。变更仅涉及 host 侧 workspace 容量与 NPU 回归测试；修复源文件及最终测试文件均已在 NPU 上验证。

## 复现与因果证据

环境固定为单机 Ascend 910_9362、TP16 × DP1、EP16、16 专家 K3、GQA DSpark 7 token、MRV1、prefix cache 关闭、`mamba_cache_mode=none`、`max_num_seqs=256`。目标使用 `FULL_DECODE_ONLY` 图回放，draft eager。vLLM 0.27.1 / `6e448d0`，vLLM-Ascend `91b905364`，CANN 9.1.0，另带此前独立的 sampler sort 内存修复；该补丁在 KDA 修复前后的服务验证中保持一致。

1. 128 并发、统一 32-token prompt 的 128 个请求成功，每个生成 256 token。
2. 随后持续 128 并发、16/32/64-token 混合 prompt，在该阶段第 7 步首次捕获 3 个未 discard 的 decode 行 count=0。这些行均有 7 个 draft token，8 个原始 sampled IDs 全为 -1。16 个 rank 的请求、count、draft 数和 discard 逐步完全一致；计数在既有异步拷贝事件完成后读取。
3. 首次复现阶段累计出现 351 个这样的 decode 行，并保存了 43 个因 NaN 无法 JSON 序列化而返回 HTTP 400 的响应。主动停止压力客户端后服务排空且健康；该轮没有 OOM，也没有把主动停服误计为崩溃。
4. 采样前现场显示 hidden states、logits 已为 NaN。关闭 target 图回放后仍能复现；进一步看到 127 条 fresh prefill 中的大批请求已从最早保留的第 7 层辅助输出开始为 NaN，而同批 spec decode 请求当时正常。
5. 完全独立于模型与 DSpark 的真实 `chunk_kda_fwd` 算子调用也复现：输入均有限且调用前后不变，Q/K/V 为 BF16、H=6、K=V=128、初始状态为 FP32 零、长度按 16/32/64 交替。127/128 条序列时输出与状态为 NaN，recompute 两种设置均失败。
6. 返回全部中间结果后，Gk、Aqk、Akk、W、Qg、Kg、H 均有限，首个异常在 U。原失败输入的 U 含 22612 个 NaN、316 个 Inf，随后传播到 V_new、输出和最终状态。

## 根因与修复

Post-WU 的 `WScratchOffset` 按每个 chunk 补齐到 `chunk_size` 后的物理跨度寻址：

```text
(((batch * value_heads + head) * total_chunks + chunk) * chunk_size + row) * key_dim
```

旧 host 分配却按未补齐的 token 数计算。127 条 16/32/64 混合序列的 token 总数是 4720，旧容量为 14,499,840 字节，正确容量为 24,969,216 字节。完整 64-token chunk 的 Cube 写入越过这一区域，可以穿过后续 FwdH 临时区并覆盖此时仍在使用的 U seed。W 在相同越界地址读回，因此 W 本身可能保持正常。

本 PR 的 [host 分配修复](https://github.com/vllm-project/vllm-ascend/blob/f27f1658669bd22f20e8b2ed539b724ef3bbb5e3/csrc/attention/chunk_kda_fwd/op_host/chunk_kda_fwd_tiling.cpp#L267-L275)将容量改为 `hChunkCount * value_heads * chunk_size * key_dim * sizeof(float)`，与实际寻址完全对齐。固定长度不是 chunk_size 整数倍、多 head 的情形也得到覆盖。新增的约 10.5 MB 是上述复现输入真实需要的空间，不是计数缓冲区扩容。

错误不要求 prefix cache 命中或启用，也不要求图回放。许多短请求同批 prefill 时，补齐长度与 token 总数差距变大，更容易暴露错误。小 batch 输出有限不能证明没有越界：它可能只覆盖后续才使用并重新初始化的临时区域。

模型 logits 为 NaN 后，随机恢复采样的比较可能无法选择有效 ID，保留 -1；随后整行有效 token 数成为 0。正常拒绝所有草稿仍应生成一个恢复 token。这里修复污染源，没有用 count clamp 或 NaN 替换掩盖症状。

## 算子验证

| 验证 | 结果 |
| --- | --- |
| 原 B127 失败输入，旧库，重复两次 | U、输出和状态均出现 NaN/Inf；输入不变 |
| 完全相同输入，仅替换修正容量的 tiling 库，重复两次 | 12 个返回张量全部有限；输入不变 |
| 新增 8 项 NPU 回归 | 全部通过；127/128 条变长及固定 80/129 token，recompute 两种模式；输出/状态对比独立 CPU 数学参考 |
| 换回旧库执行新回归 `[127-None-True]` | 失败，证明该测试可以拦住旧错误；随后恢复修复库 |
| 完整 `test_chunk_kda_aclnn.py` | 21 passed、5 skipped；A5 专用项在本次 A3 上跳过 |
| `bash format.sh ci`、`git diff --check` | 通过 |

原库 SHA256：`863b56665fcd3e0efc1e67316afea9df993c443bd6725da984afae14696e4c34`。

修复库 SHA256：`57433b68c75d4c724df1536151f7cfcb9a7ba0c55b00ef1d9aa0e2d3fb5de51f`。只增量重建 `cust_opmaster` 并原子替换验证容器的安装库；完整服务随后重启以加载新库。


## 完整服务验收

固定原 TP16/EP16、prefix cache 关闭、`max_num_seqs=256`、16 专家目标权重、GQA DSpark 7 token、target FULL_DECODE_ONLY / draft eager 配置。三轮均实际达到 128 个 running 请求。

| 阶段 | Prompt / 每请求输出 token | HTTP 200 | 输出 token | 用时 |
| --- | --- | --- | --- | --- |
| 闭合 burst，采集开启 | 32 / 256 | 128/128 | 32,768 | 72.16 秒 |
| 持续 128 并发，采集开启 | 16/32/64 / 128 | 512/512 | 65,536 | 500.38 秒 |
| 移除计数与边界采集，持续 128 并发 | 16/32/64 / 128 | 256/256 | 32,768 | 197.44 秒 |
| 总计 | | 896/896 | 131,072 | |

前两轮每个 rank 记录 718 步、73,631 个有效请求行，其中 72,991 行含 draft tokens。16 rank 的请求 ID、draft 数、discard 集合和有效 count 逐步完全相同；所有 count 为 1–8，未 discard 的 count=0、超出 `[1, draft_count+1]` 的 count 和未消费步骤均为 0。采集使用既有 CPU 拷贝完成事件，最终 RPC 排空后移除。

三轮均没有缺失或非有限的选中 token logprob，服务日志 ERROR/Traceback 为 0；每轮结束 running/waiting 都为 0，最终 `/health` 和 `/metrics` 均为 200，控制脚本退出 0。峰值 KV 使用率约 93.35%，无 preemption。16 个 worker 的进程映射已核对，均加载固定 SHA256 的修复 tiling 库。

以上统计已按逐请求响应、各阶段实际配置与 metrics、16 rank 的完整计数记录、最终健康状态交叉核对。已通过 worker 进程映射核对安装库的路径、inode 与 SHA256，确认 16 个 worker 均加载修复库。原始日志与张量保存在验证环境中。

16 专家裁剪权重的语义质量和单个 `!` 不作为故障标准。上述回归在 A3 上完成，A5 专用项已明确跳过；此前 sampler sort OOM 是另一个已独立修复的问题，其补丁在修复 KDA 前后服务中均启用。这些耗时包含调度和混合 prefill，不能用作 KDA 性能变化的对照。

## PR CI 补充

PR 初次 CI 的 mypy 在新增测试中报告 `reference.final_state` 可能为 None。已增加显式非空断言，提交 `f27f1658669bd22f20e8b2ed539b724ef3bbb5e3` 并推送同一 PR；当前 PR head 为该提交，生产算子修复仍为 `8ddf72b3b697c29b8b0cd98401dd396fa319fd26`。`select-tests` 因前置检查失败而跳过 coverage 推荐并失败。

断言加入后，NPU 上重跑新增回归 8 passed；本地对该文件的 Python 3.10/3.11/3.12 mypy 均通过（follow-imports skip，且本地缺少 torch/pytest，使用 ignore-missing-imports），`bash format.sh ci` 再次通过。[当前 head 的 E2E workflow](https://github.com/vllm-project/vllm-ascend/actions/runs/34253759987) 已触发，发布此评论时 pre-commit 正在排队；DCO 与 PR create action 已通过。尚未将这一轮 CI 计为完成的通过证据。生产源码未改变，因此前述完整服务结果仍对应当前 PR 的算子修复。
